packages/gatsby/src/utils/webpack.config.js

```diff
@@ -468,13 +468,30 @@ module.exports = async (
   }
 
   if (stage === `build-javascript`) {
+    const componentsCount = store.getState().components.size
+
     config.optimization = {
       runtimeChunk: {
         name: `webpack-runtime`,
       },
       splitChunks: {
         name: false,
+        chunks: `all`,
         cacheGroups: {
+          default: false,
+          vendors: false,
+          commons: {
+            name: `commons`,
+            chunks: `all`,
+            // if a chunk is used more than half the components count,
+            // we can assume it's pretty global
+            minChunks: componentsCount > 2 ? componentsCount * 0.5 : 2,
```

Comment:
More than half components count is a nice metric 😅

Reply:
Does `store.getState().components.size` include pages?

Reply:
I need components as they are the ones that create chunks. We can have 1000 of pages that only use 1 component so we need to check components and not pages.

Reply:
Hmm, fair enough

Reply:
Crazy idea: GuessJS for webpack config :D

Reply:
Note I've opened a related issue where you'll find links to what has been done recently in NextJS which has a new "granularChunks" config: https://github.com/gatsbyjs/gatsby/issues/16661

Reply:
Yeah so the idea is to go on an even more granular level, like nextjs but I wanted something I could get in fairly easy without having to check x amount of sites.